test/acceptance/tests/hcp/hcp_test.go

```diff
@@ -1,124 +1,119 @@
-	randomSuffix := strings.ToLower(random.UniqueId())
-	tfVars := suite.Config().TFVars()
-	tfVars["suffix"] = randomSuffix
-	tfOptions := &terraform.Options{
-		TerraformDir: "./terraform/hcp-install",
-		Vars:         tfVars,
-		NoColor:      true,
-	}
-	terraformOptions := terraform.WithDefaultRetryableErrors(t, tfOptions)
+	// read the configuration from the setup-terraform dir.
+	var cfg HCPTestConfig
+	require.NoError(t, config.UnmarshalTF(setupDir, &cfg))
+
+	// generate input variables to the test terraform using the config.
+	ignoreVars := []string{"ecs_cluster_1_arn", "ecs_cluster_2_arn"}
+	tfVars := config.TFVars(cfg, ignoreVars...)
 
+	consulClient, consulState, err := consulClient(t, cfg.ConsulAddr, cfg.ConsulToken)
+	require.NoError(t, err)
 	defer func() {
-		if suite.Config().NoCleanupOnFailure && t.Failed() {
-			logger.Log(t, "skipping resource cleanup because -no-cleanup-on-failure=true")
-		} else {
-			terraform.Destroy(t, terraformOptions)
+		if err := restoreConsulState(t, consulClient, consulState); err != nil {
+			t.Log(err)
 		}
 	}()
-	terraform.InitAndApply(t, terraformOptions)
 
-	outputs := terraform.OutputAll(t, &terraform.Options{
-		TerraformDir: "./terraform/hcp-install",
-		NoColor:      true,
-		Logger:       terratestLogger.Discard,
-	})
+	randomSuffix := strings.ToLower(random.UniqueId())
 
-	cfg := api.DefaultConfig()
-	cfg.Address = outputs["hcp_public_endpoint"].(string)
-	cfg.Token = outputs["token"].(string)
-	consulClient, err := api.NewClient(cfg)
-	require.NoError(t, err)
+	clientTask := &helpers.MeshTask{
+		Name:         fmt.Sprintf("test_client_%s", randomSuffix),
+		Partition:    "default",
+		Namespace:    "default",
+		ConsulClient: consulClient,
+		Region:       cfg.Region,
+		ClusterARN:   cfg.ECSClusterARN,
+	}
+	serverTask := &helpers.MeshTask{
+		Name:         fmt.Sprintf("test_server_%s", randomSuffix),
+		Partition:    "default",
+		Namespace:    "default",
+		ConsulClient: consulClient,
+		Region:       cfg.Region,
+		ClusterARN:   cfg.ECSClusterARN,
+	}
 
-	serverServiceName := fmt.Sprintf("test_server_%s", randomSuffix)
-	clientServiceName := fmt.Sprintf("test_client_%s", randomSuffix)
+	tfVars["suffix"] = randomSuffix
+	terraformOptions, _ := terraformInitAndApply(t, "./terraform/hcp-install", tfVars)
+	defer terraformDestroy(t, terraformOptions, suite.Config().NoCleanupOnFailure)
 
 	// Wait for both tasks to be registered in Consul.
-	logger.Log(t, "checking if services are registered")
-	retry.RunWith(&retry.Timer{Timeout: 6 * time.Minute, Wait: 20 * time.Second}, t, func(r *retry.R) {
-		services, _, err := consulClient.Catalog().Services(nil)
-		r.Check(err)
-		logger.Logf(t, "Consul services: %v", services)
-		require.Contains(r, services, serverServiceName)
-		require.Contains(r, services, clientServiceName)
+	logger.Log(t, "waiting for services to register with consul")
+	retry.RunWith(timeouts["registration"], t, func(r *retry.R) {
+		require.True(r, clientTask.Registered())
```

Comment:
The code reads much better after this refactoring. Great work!

Reply:
Thanks :)